wellcomeml/ml/clustering.py

```diff
@@ -292,6 +304,74 @@ def optimise(self, X, param_grid, n_cluster_range=None, max_noise=0.2,
 
         return best_params
 
+    def save(self, folder, components='all', create_folder=True):
+        """
+        Saves the different steps of the pipeline
+
+        Args:
+            folder(str): path to folder
+            components(list or 'all'): List of components to save. Options are: 'embbedded_points',
+            'reduced_points', 'vectorizer', 'reducer', and 'clustering_model'. By default, loads
+            'all' (you can get all components by listing the class param
+             TextClustering.components)
+
+        """
+        if create_folder:
+            os.makedirs(folder, exist_ok=True)
+
+        if components == 'all' or 'embedded_points' in components:
+            np.save(os.path.join(folder, 'embedded_points.npy'), self.embedded_points)
+
+        if components == 'all' or 'reduced_points' in components:
+            np.save(os.path.join(folder, 'reduced_points.npy'), self.reduced_points)
+
+        if components == 'all' or 'vectorizer' in components:
+            with open(os.path.join(folder, 'vectorizer.pkl'), 'wb') as f:
+                pickle.dump(self.vectorizer, f)
+
+        if components == 'all' or 'reducer' in components:
+            with open(os.path.join(folder, 'reducer.pkl'), 'wb') as f:
+                pickle.dump(self.reducer_class, f)
+
+        if components == 'all' or 'clustering_model' in components:
+            with open(os.path.join(folder, 'clustering.pkl'), 'wb') as f:
+                pickle.dump(self.clustering_class, f)
+
+    def load(self, folder, components='all'):
+        """
+        Loads the different steps of the pipeline
+
+        Args:
+            folder(str): path to folder
+            components(list or 'all'): List of components to load. Options are: 'embbedded_points',
+            'reduced_points', 'vectorizer', 'reducer', and 'clustering_model'. By default, loads
+            'all' (you can get all components by listing the class param
+             TextClustering.components)
+
+        """
+
+        if components == 'all' or 'embedded_points' in components:
+            self.embedded_points = np.load(os.path.join(folder, 'embedded_points.npy'),
```

Comment:
there is some duplication about saving and loading the paths. you can create them in a variable that stays in the class and reuse `self.embedded_points_path = xxx` possibly in init or something

Reply:
As I pushed the code I thought "Is Nick going to comment on code duplication?" 😆 . 

That's a very fair point, will address.